### PR TITLE
NETWORKING: Add SSL Handler before other Handlers

### LINKED ISSUE #33998: [CI] SSL setup fails in tests
Build: https://elasticsearch-ci.elastic.co/job/elastic+elasticsearch+master+multijob-unix-compatibility/os=opensuse/20/consoleFull

Couldn't reproduce this, with several local runs yet:
```
./gradlew :x-pack:plugin:security:test \
  -Dtests.seed=8F67CB2DAC5BD0EE \
  -Dtests.class=org.elasticsearch.integration.ShrinkIndexWithSecurityTests \
  -Dtests.security.manager=true \
  -Dtests.locale=en-US \
  -Dtests.timezone=Etc/UTC \
  -Dcompiler.java=10 \
  -Druntime.java=8
```

The failure appears to be happening already at test setup time:
```
14:48:55   2> WARNING: Uncaught exception in thread: Thread[elasticsearch[node_sm2][generic][T#4],5,TGRP-ShrinkIndexWithSecurityTests]
14:48:55   2> java.lang.AssertionError: Must have SslHandler
14:48:55   1> [2018-09-24T03:48:54,584][INFO ][o.e.n.Node               ] [suite] stopped
14:48:55   1> [2018-09-24T03:48:54,585][INFO ][o.e.n.Node               ] [suite] closing ...
14:48:55   1> [2018-09-24T03:48:54,586][INFO ][o.e.n.Node               ] [suite] closed
14:48:55   1> [2018-09-24T03:48:54,631][INFO ][o.e.n.Node               ] [suite] stopping ...
14:48:55   2> 	at __randomizedtesting.SeedInfo.seed([8F67CB2DAC5BD0EE]:0)
14:48:55   2> 	at org.elasticsearch.xpack.security.transport.SSLEngineUtils.getSSLEngine(SSLEngineUtils.java:62)
14:48:55   2> 	at org.elasticsearch.xpack.security.transport.SSLEngineUtils.extractClientCertificates(SSLEngineUtils.java:39)
14:48:55   2> 	at org.elasticsearch.xpack.security.transport.ServerTransportFilter$NodeProfile.inbound(ServerTransportFilter.java:113)
14:48:55   2> 	at org.elasticsearch.xpack.security.transport.SecurityServerTransportInterceptor$ProfileSecuredRequestHandler.messageReceived(SecurityServerTransportInterceptor.java:307)
14:48:55   2> 	at org.elasticsearch.transport.RequestHandlerRegistry.processMessageReceived(RequestHandlerRegistry.java:63)
14:48:55   2> 	at org.elasticsearch.transport.TcpTransport$RequestHandler.doRun(TcpTransport.java:1448)
14:48:55   2> 	at org.elasticsearch.common.util.concurrent.ThreadContext$ContextPreservingAbstractRunnable.doRun(ThreadContext.java:723)
14:48:55   2> 	at org.elasticsearch.common.util.concurrent.AbstractRunnable.run(AbstractRunnable.java:37)
14:48:55   2> 	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
14:48:55   2> 	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
```

### PR DESCRIPTION
**Pull request by original-brownbear**
* The only way to run into the issue in #33998 is for `Netty4MessageChannelHandler`
to be in the pipeline while the SslHandler is not. Adding the SslHandler before any
other handlers should ensure correct ordering here even when we handle upstream events
in our own thread pool
* Relates #33998 
   * I would like to not close #33998 yet since it looks to me like this is a Netty bug (though I couldn't reproduce it in isolation yet) that somehow causes the `Netty4MessageChannelHandler` to be invoked before the initialize call returns (though the first "UGLY" point in https://github.com/netty/netty/wiki/Thread-model could be the culprit here ... if we don't see the error anymore with this change we could look into opening an issue with Netty I think).

### CODE REVIEW

**Review by Tim-Brooks — CHANGES_REQUESTED** [review]
Thanks for picking this up. I missed the original issue.

I'm suspicious that this does not solve the issue. `initChannel` should not be called concurrently with the channel being read from. Additionally, `SslHandler` is necessary to have received a message from a channel. We need to handshake + decode in order to successfully get to the `TcpTransport`.

I suspect the actual issue is that the channel is being closed concurrently with a request being handled. When a channel is closed, netty removes the handlers. 

`DefaultChannelPipeline`

```
    private void destroyUp(AbstractChannelHandlerContext ctx, boolean inEventLoop) {
        final Thread currentThread = Thread.currentThread();
        final AbstractChannelHandlerContext tail = this.tail;
        for (;;) {
            if (ctx == tail) {
                destroyDown(currentThread, tail.prev, inEventLoop);
                break;
            }

            final EventExecutor executor = ctx.executor();
            if (!inEventLoop && !executor.inEventLoop(currentThread)) {
                final AbstractChannelHandlerContext finalCtx = ctx;
                executor.execute(new Runnable() {
                    @Override
                    public void run() {
                        destroyUp(finalCtx, true);
                    }
                });
                break;
            }

            ctx = ctx.next;
            inEventLoop = false;
        }
    }
```

I think what we need to do is check whether the channel is closed when the `SslHandler` is null. If the channel is open, then throw the assertion error. That is not excepted. Otherwise maybe throw some type of normal exception to indicate that the request cannot be processed? I'll ping the @elastic/es-security team here to see what they think. What should we do if we cannot access the SSLEngine as the channel if closed already?

**Comment by original-brownbear** [comment]
> `initChannel` should not be called concurrently with the channel being read from. Additionally, SslHandler is necessary to have received a message from a channel. We need to handshake + decode in order to successfully get to the TcpTransport.

I think the issue isn't that the handler isn't there but rather that there may be a concurrency issue in the lookup of the handler (the thread failing the assertion in the report is not a Netty eventloop thread, see the thread model wiki page I linked where a race condition in this code is mentioned).

> I think what we need to do is check whether the channel is closed when the SslHandler is null. If the channel is open, then throw the assertion error. That is not excepted. Otherwise maybe throw some type of normal exception to indicate that the request cannot be processed?

Yea, this seems like a good idea! We are currently just checking if the channel is open before the we go into the method failing the assertion but since we're not in a Netty thread here we probably have the same concurrency issue here as well. I'd just throw an `io.netty.channel.ChannelException` in this case.
That said, when debugging this I actually tried if I could reproduce this by closing the channel concurrently (tried from within the Eventloop as well as from just our threadpool ... the `SslHandler` could still be looked up after a few seconds but maybe if the connection is closed from a broken pipe concurrently it's deregistered immediately?).
Either way, can't hurt to rule this out as well so I added https://github.com/elastic/elasticsearch/pull/34636/commits/b92fe67d88812a63b8dab639143c6863af16dbc1 :)

**Comment by jaymode** [comment]
> Otherwise maybe throw some type of normal exception to indicate that the request cannot be processed?

+1

> see the thread model wiki page I linked where a race condition in this code is mentioned

I've looked at this before and I've always thought this described the Netty 3.x thread model and its issues that they wanted to fix in 4.0. The history of that page shows most of it was written 5 years ago and recent changes are very minor wording fixes, https://github.com/netty/netty/wiki/Thread-model/_history. I'm not sure we should base things off of this unless we can confirm that these issues still apply.

**Comment by tbrooks8** [comment]
> I've looked at this before and I've always thought this described the Netty 3.x thread model and its issues that they wanted to fix in 4.0. 

This is referencing 3.x. All inbound and outbound events happen on the same thread now. 

> I think the issue isn't that the handler isn't there but rather that there may be a concurrency issue in the lookup of the handler

I'm open to some complicated concurrency issue. But looking up a handler only touches `final` or `volatile` fields. If we agree that the handler is there (in order to have completed a handshake), we should be able to access it.

Anyway, I'm good with both changes to be safe. If Jay is okay with the `ChannelException` as the notification mechanism. And one comment that I just added.

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/transport/netty4/SecurityNetty4Transport.java [hunk 1]
```diff
@@ -157,11 +157,11 @@ public SslChannelInitializer(String name, SSLConfiguration configuration) {
 
         @Override
         protected void initChannel(Channel ch) throws Exception {
-            super.initChannel(ch);
             SSLEngine serverEngine = sslService.createSSLEngine(configuration, null, -1);
             serverEngine.setUseClientMode(false);
             final SslHandler sslHandler = new SslHandler(serverEngine);
             ch.pipeline().addFirst("sslhandler", sslHandler);
+            super.initChannel(ch);
```
Comment (tbrooks8) [inline x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/transport/netty4/SecurityNetty4Transport.java]: Can you add an assertion post `super` call that the ssl handler is still first? That way a future change does not accidentally mess up necessary ordering on handlers.

**Comment by jaymode** [comment]
> If Jay is okay with the ChannelException as the notification mechanism.

I am good with it

**Review by Tim-Brooks — APPROVED** [review]
LGTM

**Comment by original-brownbear** [comment]
@tbrooks8 thanks!

**Pull request merged**